lib/oli/delivery/metrics.ex

```diff
@@ -857,4 +842,59 @@ defmodule Oli.Delivery.Metrics do
 
     Repo.one(query)
   end
+
+  # Given a list of ContainedPage records, return a map of
+  # page ids to a list of container ids (their ancestor containers)
+  # that contain that page
+  #
+  # For instance, given the following ContainedPage records:
+  #
+  #  %ContainedPage{container_id: 1, page_id: 10}
+  #  %ContainedPage{container_id: 1, page_id: 11}
+  #  %ContainedPage{container_id: 2, page_id: 10}
+  #
+  # This function will return:
+  # %{10 => [1, 2], 11 => [1]}
+  defp page_to_parent_containers_map(contained_pages) do
+    Enum.reduce(contained_pages, %{},
+      fn %ContainedPage{container_id: container_id, page_id: page_id}, inverted_cp_index ->
+        case Map.get(inverted_cp_index, page_id) do
+          nil -> Map.put(inverted_cp_index, page_id, [container_id])
+          container_ids -> Map.put(inverted_cp_index, page_id, [container_id | container_ids])
+        end
+      end)
+  end
+
+  # Given a list of ContainedPage records, return a map of container ids to a tuple of
+  # correct and total values, initialized to {0.0, 0.0}
+  defp init_container_totals(contained_pages) do
+     Enum.reduce(contained_pages, %{}, fn %ContainedPage{container_id: container_id}, map ->
+      Map.put(map, container_id, {0.0, 0.0})
+    end)
+  end
+
+  # Given a list of {page_id, correct, total} tuples, and a list of
+  # ContainedPage records, return a map of container ids to a tuple of correct and total values,
+  # where the container totals are the sum of the page totals for all pages contained in that container.
+  defp bucket_into_container_totals(page_totals, contained_pages) do
+
+    inverted_cp_index = page_to_parent_containers_map(contained_pages)
+    container_totals = init_container_totals(contained_pages)
+
+    Enum.reduce(page_totals, container_totals, fn {page_id, correct, total}, map ->
+      container_ids = Map.get(inverted_cp_index, page_id)
+      Enum.reduce(container_ids, map, fn container_id, map ->
```

Comment:
I'm running into an error here when testing with the chemistry course, I believe it is related to `container_ids` being null, possibly from detatched pages?

```[error] Task #PID<0.30102.0> started from #PID<0.30094.0> terminating
** (Protocol.UndefinedError) protocol Enumerable not implemented for nil of type Atom
    (elixir 1.14.0) lib/enum.ex:1: Enumerable.impl_for!/1
    (elixir 1.14.0) lib/enum.ex:166: Enumerable.reduce/3
    (elixir 1.14.0) lib/enum.ex:4307: Enum.reduce/3
    (elixir 1.14.0) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
    (oli 0.24.3) lib/oli/delivery/metrics.ex:886: Oli.Delivery.Metrics.bucket_into_container_totals/2
    (oli 0.24.3) lib/oli_web/live/delivery/instructor_dashboard/instructor_dashboard_live.ex:599: anonymous fn/2 in OliWeb.Delivery.InstructorDashboard.InstructorDashboardLive.async_calculate_proficiency/1
    (elixir 1.14.0) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (elixir 1.14.0) lib/task/supervised.ex:34: Task.Supervised.reply/4
    (stdlib 3.15) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Function: #Function<0.102994098/0 in OliWeb.Delivery.InstructorDashboard.InstructorDashboardLive.async_calculate_proficiency/1>
    Args: []
[error] an exception was raised:
    ** (Protocol.UndefinedError) protocol Enumerable not implemented for nil of type Atom
        (elixir 1.14.0) lib/enum.ex:1: Enumerable.impl_for!/1
        (elixir 1.14.0) lib/enum.ex:166: Enumerable.reduce/3
        (elixir 1.14.0) lib/enum.ex:4307: Enum.reduce/3
        (elixir 1.14.0) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
        (oli 0.24.3) lib/oli/delivery/metrics.ex:886: Oli.Delivery.Metrics.bucket_into_container_totals/2
        (oli 0.24.3) lib/oli_web/live/delivery/instructor_dashboard/instructor_dashboard_live.ex:599: anonymous fn/2 in OliWeb.Delivery.InstructorDashboard.InstructorDashboardLive.async_calculate_proficiency/1
        (elixir 1.14.0) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
        (elixir 1.14.0) lib/task/supervised.ex:34: Task.Supervised.reply/4
        (stdlib 3.15) proc_lib.erl:226: :proc_lib.init_p_do_apply/3```

Reply:
Indeed adding a case statement around this code fixes the issue, however I am not certain this is what we want here or not
```
      case container_ids do
        nil -> map
        _ ->
          Enum.reduce(container_ids, map, fn container_id, map ->
            {current_correct, current_total} = Map.get(map, container_id)
            Map.put(map, container_id, {current_correct + correct, current_total + total})
          end)
      end
```

Reply:
Ah, yes, great catch.  That is exactly what is happening here.  You're solution makes sense since an unordered page has no ancestor containers - we simply have to ignore it.